File: examples/contrib/set_partition.py
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Set partition problem in Google CP Solver.

Problem formulation from
http://www.koalog.com/resources/samples/PartitionProblem.java.html
'''
 This is a partition problem.
 Given the set S = {1, 2, ..., n},
 it consists in finding two sets A and B such that:

   A U B = S,
   |A| = |B|,
   sum(A) = sum(B),
   sum_squares(A) = sum_squares(B)

'''

This model uses a binary matrix to represent the sets.


Also, compare with other models which uses var sets:
* MiniZinc: http://www.hakank.org/minizinc/set_partition.mzn
* Gecode/R: http://www.hakank.org/gecode_r/set_partition.rb
* Comet: http://hakank.org/comet/set_partition.co
* Gecode: http://hakank.org/gecode/set_partition.cpp
* ECLiPSe: http://hakank.org/eclipse/set_partition.ecl
* SICStus: http://hakank.org/sicstus/set_partition.pl

This model was created by Hakan Kjellerstrand (hakank@gmail.com)
Also see my other Google CP Solver models:
http://www.hakank.org/google_or_tools/
"""

import sys

from ortools.constraint_solver.python import constraint_solver as cp


#
# Partition the sets (binary matrix representation).
#
def partition_sets(x, num_sets, n):
    solver = list(x.values())[0].solver

    for i in range(num_sets):
        for j in range(num_sets):
            if i != j:
                b = solver.sum([x[i, k] * x[j, k] for k in range(n)])
                solver.add(b == 0)

    # ensure that all integers is in
    # (exactly) one partition
    b = [x[i, j] for i in range(num_sets) for j in range(n)]
    solver.add(solver.sum(b) == n)


def main(n=16, num_sets=2):

    # Create the solver.
    solver = cp.Solver("Set partition")

    #
    # data
    #
    print("n:", n)
    print("num_sets:", num_sets)
    print()

    # Check sizes
    assert n % num_sets == 0, "Equal sets is not possible."

    #
    # variables
    #

    # the set
    a = {}
    for i in range(num_sets):
        for j in range(n):
            a[i, j] = solver.new_int_var(0, 1, "a[%i,%i]" % (i, j))

    a_flat = [a[i, j] for i in range(num_sets) for j in range(n)]

    #
    # constraints
    #

    # partition set
    partition_sets(a, num_sets, n)

    for i in range(num_sets):
        for j in range(i, num_sets):

            # same cardinality
            solver.add(
                solver.sum([a[i, k] for k in range(n)])
                == solver.sum([a[j, k] for k in range(n)])
            )

            # same sum
            solver.add(
                solver.sum([k * a[i, k] for k in range(n)])
                == solver.sum([k * a[j, k] for k in range(n)])
            )

            # same sum squared
            solver.add(
                solver.sum([(k * a[i, k]) * (k * a[i, k]) for k in range(n)])
                == solver.sum([(k * a[j, k]) * (k * a[j, k]) for k in range(n)])
            )

    # symmetry breaking for num_sets == 2
    if num_sets == 2:
        solver.add(a[0, 0] == 1)

    #
    # search and result
    #
    db = solver.phase(
        a_flat,
        cp.IntVarStrategy.INT_VAR_DEFAULT,
        cp.IntValueStrategy.INT_VALUE_DEFAULT,
    )

    solver.new_search(db)

    num_solutions = 0
    while solver.next_solution():
        a_val = {}
        for i in range(num_sets):
            for j in range(n):
                a_val[i, j] = a[i, j].value()

        sq = sum([(j + 1) * a_val[0, j] for j in range(n)])
        print("sums:", sq)
        sq2 = sum([((j + 1) * a_val[0, j]) ** 2 for j in range(n)])
        print("sums squared:", sq2)

        for i in range(num_sets):
            if sum([a_val[i, j] for j in range(n)]):
                print(i + 1, ":", end=" ")
                for j in range(n):
                    if a_val[i, j] == 1:
                        print(j + 1, end=" ")
                print()

        print()
        num_solutions += 1

    solver.end_search()

    print()
    print("num_solutions:", num_solutions)
    print("failures:", solver.num_failures)
    print("branches:", solver.num_branches)
    print("WallTime:", solver.wall_time_ms)


n = 16
num_sets = 2
if __name__ == "__main__":
    if len(sys.argv) > 1:
        n = int(sys.argv[1])
    if len(sys.argv) > 2:
        num_sets = int(sys.argv[2])

    main(n, num_sets)

File: examples/contrib/set_partition.cs
//
// Copyright 2012 Hakan Kjellerstrand
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Diagnostics;
using Google.OrTools.ConstraintSolver;

public class SetPartition
{
    //
    // Partition the sets (binary matrix representation).
    //
    public static void partition_sets(Solver solver, IntVar[,] x, int num_sets, int n)
    {
        for (int i = 0; i < num_sets; i++)
        {
            for (int j = 0; j < num_sets; j++)
            {
                if (i != j)
                {
                    // b = solver.Sum([x[i,k]*x[j,k] for k in range(n)]);
                    // solver.Add(b == 0);
                    solver.Add((from k in Enumerable.Range(0, n) select(x[i, k] * x[j, k])).ToArray().Sum() == 0);
                }
            }
        }

        // ensure that all integers is in
        // (exactly) one partition
        solver.Add(
            (from i in Enumerable.Range(0, num_sets) from j in Enumerable.Range(0, n) select x[i, j]).ToArray().Sum() ==
            n);
    }

    /**
     *
     * Set partition problem.
     *
     * Problem formulation from
     * http://www.koalog.com/resources/samples/PartitionProblem.java.html
     * """
     * This is a partition problem.
     * Given the set S = {1, 2, ..., n},
     * it consists in finding two sets A and B such that:
     *
     *  A U B = S,
     *  |A| = |B|,
     *  sum(A) = sum(B),
     *  sum_squares(A) = sum_squares(B)
     *
     * """
     *
     * This model uses a binary matrix to represent the sets.
     *
     *
     * Also see http://www.hakank.org/or-tools/set_partition.py
     *
     */
    private static void Solve(int n = 16, int num_sets = 2)
    {
        Solver solver = new Solver("SetPartition");

        Console.WriteLine("n: {0}", n);
        Console.WriteLine("num_sets: {0}", num_sets);

        IEnumerable<int> Sets = Enumerable.Range(0, num_sets);
        IEnumerable<int> NRange = Enumerable.Range(0, n);

        //
        // Decision variables
        //
        IntVar[,] a = solver.MakeIntVarMatrix(num_sets, n, 0, 1, "a");
        IntVar[] a_flat = a.Flatten();

        //
        // Constraints
        //

        // partition set
        partition_sets(solver, a, num_sets, n);

        foreach (int i in Sets)
        {
            foreach (int j in Sets)
            {
                // same cardinality
                solver.Add((from k in NRange select a[i, k]).ToArray().Sum() ==
                           (from k in NRange select a[j, k]).ToArray().Sum());

                // same sum
                solver.Add((from k in NRange select(k * a[i, k])).ToArray().Sum() ==
                           (from k in NRange select(k * a[j, k])).ToArray().Sum());

                // same sum squared
                solver.Add((from k in NRange select(k * a[i, k] * k * a[i, k])).ToArray().Sum() ==
                           (from k in NRange select(k * a[j, k] * k * a[j, k])).ToArray().Sum());
            }
        }

        // symmetry breaking for num_sets == 2
        if (num_sets == 2)
        {
            solver.Add(a[0, 0] == 1);
        }

        //
        // Search
        //
        DecisionBuilder db = solver.MakePhase(a_flat, Solver.INT_VAR_DEFAULT, Solver.INT_VALUE_DEFAULT);

        solver.NewSearch(db);

        while (solver.NextSolution())
        {
            int[,] a_val = new int[num_sets, n];
            foreach (int i in Sets)
            {
                foreach (int j in NRange)
                {
                    a_val[i, j] = (int)a[i, j].Value();
                }
            }
            Console.WriteLine("sums: {0}", (from j in NRange select(j + 1) * a_val[0, j]).ToArray().Sum());

            Console.WriteLine("sums squared: {0}",
                              (from j in NRange select(int) Math.Pow((j + 1) * a_val[0, j], 2)).ToArray().Sum());

            // Show the numbers in each set
            foreach (int i in Sets)
            {
                if ((from j in NRange select a_val[i, j]).ToArray().Sum() > 0)
                {
                    Console.Write(i + 1 + ": ");
                    foreach (int j in NRange)
                    {
                        if (a_val[i, j] == 1)
                        {
                            Console.Write((j + 1) + " ");
                        }
                    }
                    Console.WriteLine();
                }
            }
            Console.WriteLine();
        }

        Console.WriteLine("\nSolutions: {0}", solver.Solutions());
        Console.WriteLine("WallTime: {0}ms", solver.WallTime());
        Console.WriteLine("Failures: {0}", solver.Failures());
        Console.WriteLine("Branches: {0} ", solver.Branches());

        solver.EndSearch();
    }

    public static void Main(String[] args)
    {
        int n = 16;
        int num_sets = 2;

        if (args.Length > 0)
        {
            n = Convert.ToInt32(args[0]);
        }

        if (args.Length > 1)
        {
            num_sets = Convert.ToInt32(args[1]);
        }

        if (n % num_sets == 0)
        {
            Solve(n, num_sets);
        }
        else
        {
            Console.WriteLine("n {0} num_sets {1}: Equal sets is not possible!", n, num_sets);
        }
    }
}
